controllers/upgrade_controller.go

```diff
@@ -152,6 +153,13 @@ func (r *UpgradeReconciler) BuildState(ctx context.Context) (*upgrade.ClusterUpg
 
 	r.Log.V(consts.LogLevelDebug).Info("Got driver daemon sets", "length", len(daemonSets))
 
+	for _, ds := range daemonSets {
```

Comment:
as discussed offline, the DS status may not yet be updated by k8s daemonset controller. so we may be called after pod delete but before ds got a chance to update its status, leading to the same issue.

i think we should rely on ds status DesiredNumberScheduled (which just maps to the number of nodes that should run a ds pod) as this is not expected to change (scale up / down) in the middle of an active update flow.

then we can list the pods and count the ones associated with daemonset (have owner reference) and dont have deletion timestap set.

if the two are the same it means that pods that were deleted got rescheduled by the daemon. so we sort of do a similar logic to k8s daemonset controller.

https://github.com/kubernetes/kubernetes/blob/a1c8e9386af844757333733714fa1757489735b3/pkg/controller/daemon/daemon_controller.go#L755